src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1948,6 +2117,126 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
 	if (!is_target_path)
 		goto org_func;
 
+	/* Always rely on fuse for open() to avoid a fake fd */
+	if (compatible_mode) {
+		struct ht_fd *fd_ht_obj = NULL;
+		int           fd_fake   = 0;
+
+		if (two_args)
+			fd_kernel = real_open(pathname, oflags);
+		else
+			fd_kernel = real_open(pathname, oflags, mode);
+		/* dfuse fails for some reason */
+		if (fd_kernel < 0)
+			goto out_compatible;
+
+		/* query file object through ioctl() */
+		rc = fetch_dfs_cont_file_obj_with_fd(fd_kernel, dfs_mt, &dfs_obj);
+		if (rc != 0 && rc != EISDIR) {
+			DS_WARN(rc, "fetch_dfs_cont_file_obj_with_fd() failed");
+			goto out_compatible;
+		}
+
+		/* Need to create a fake fd and associate with fd_kernel */
+		atomic_fetch_add_relaxed(&num_open, 1);
+
+		/* regular file */
+		if (rc == 0) {
+			rc = find_next_available_fd(NULL, &idx_fd);
+			if (rc)
+				goto out_compatible_release;
+			file_list[idx_fd]->dfs_mt    = dfs_mt;
+			file_list[idx_fd]->file      = dfs_obj;
+			file_list[idx_fd]->parent    = parent;
+			file_list[idx_fd]->st_ino    = FAKE_ST_INO(full_path);
+			file_list[idx_fd]->idx_mmap  = -1;
+			file_list[idx_fd]->open_flag = oflags;
+			file_list[idx_fd]->offset    = 0;
+			if (strncmp(full_path, "/", 2) == 0)
+				D_STRNDUP(file_list[idx_fd]->path, dfs_mt->fs_root, DFS_MAX_PATH);
+			else
+				D_ASPRINTF(file_list[idx_fd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (file_list[idx_fd]->path == NULL) {
+				free_fd(idx_fd, false);
+				goto out_compatible;
+			}
+			strncpy(file_list[idx_fd]->item_name, item_name, DFS_MAX_NAME);
+			fd_fake = idx_fd + FD_FILE_BASE;
+		}
+
+		/* directory */
+		if (rc == EISDIR) {
```

Comment:
can you explain how EISDIR is returned in this case and why it can't be handled like files and is done on error?

Reply:
ioctl(fd, DFUSE_IOCTL_IL_DSIZE, &hsd_reply);
fails with "EISDIR" if fd is a file descriptor opened for a directory. 

https://github.com/daos-stack/daos/blob/master/src/client/dfuse/ops/ioctl.c#L441-L442

@ashleypittman Do you know why a dir is handled differently? Thank you!

Reply:
It'll be because of https://github.com/daos-stack/daos/blob/master/src/client/dfuse/ops/ioctl.c#L441-L442

calling the ioctl on a directory doesn't make sense for `ioil` but does in this case so we could probably remove that check, the ioctl is also used by other things like libduns and the daos command though so we'd need to verify they weren't affected.

Reply:
@ashleypittman Thank you very much! If I remove this check, what test cases we should run in CI for other affected component? Is "Features: pil4dfs" currently used sufficient?

Reply:
For now, I'll try and work through the implications tomorrow and will update you with anything else that may be required.

Reply:
Thank you very much!

Reply:
@ashleypittman 
https://github.com/daos-stack/daos/blob/master/src/client/dfuse/ops/ioctl.c#L230
For a directory, oh->doh_obj is NULL. Consequently dfs_obj_local2global() returns EINVAL.

Reply:
For a directory try `oh->doh_ie->ie_obj`

Reply:
oh->doh_ie->ie_obj does help.
dfs_obj_local2global() still returns EINVAL because glob->iov_buf is NULL and glob->iov_buf_len is zero.
https://github.com/daos-stack/daos/blob/master/src/client/dfs/obj.c#L574-L575

Reply:
Are you sure that errors not coming from https://github.com/daos-stack/daos/blob/master/src/client/dfs/obj.c#L566

the local2global functions are intentionally called twice, the first time just fetches the size, the second call is after an appropriate sized buffer has been allocated, then the second call should fill in the buffer.

Reply:
I already revised the code in this line.
if (obj == NULL || !S_ISREG(obj->mode))
to
if (obj == NULL || (!S_ISREG(obj->mode) **&& !S_ISDIR(obj->mode)**))

Reply:
That change isn't in the PR but as you've put it there the brackets are incorrect.

Reply:
Do you have a resolution for this?